core/statix.multilang/src/main/java/mb/statix/multilang/SharedPieProvider.java

```diff
@@ -0,0 +1,58 @@
+package mb.statix.multilang;
+
+import dagger.Lazy;
+import mb.pie.api.ExecException;
+import mb.pie.api.MixedSession;
+import mb.pie.api.Pie;
+import mb.resource.hierarchical.ResourcePath;
+import mb.spoofax.core.language.LanguageScope;
+import mb.spoofax.core.pie.PieProvider;
+import mb.statix.multilang.metadata.ContextId;
+import mb.statix.multilang.metadata.ContextPieManager;
+import mb.statix.multilang.metadata.LanguageId;
+import mb.statix.multilang.pie.config.SmlBuildContextConfiguration;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Set;
+
+@LanguageScope
+public class SharedPieProvider implements PieProvider {
+
+    private final Pie languagePie;
+    private final LanguageId languageId;
+    private final Lazy<ContextPieManager> contextPieManager;
+    private final SmlBuildContextConfiguration buildContextConfiguration;
+
+    @Inject public SharedPieProvider(
+        @Named("prototype") Pie languagePie,
+        LanguageId languageId,
+        Lazy<ContextPieManager> contextPieManager,
+        SmlBuildContextConfiguration buildContextConfiguration
+    ) {
+        this.languagePie = languagePie;
+        this.languageId = languageId;
+        this.contextPieManager = contextPieManager;
+        this.buildContextConfiguration = buildContextConfiguration;
+    }
+
+    @Override
+    public Pie getPie(@Nullable ResourcePath projectDir) {
```

Comment:
I really dislike that a project path is used to create a PIE instance. Getting a PIE instance should be decoupled from project paths, or any kinds of paths. Only tasks of a language that require a project path should have access to it.